File: src/main/java/dev/strawn/frugal/FrugalLexer.java
package dev.strawn.frugal;

import com.intellij.lexer.LexerBase;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class FrugalLexer extends LexerBase {
    
    private CharSequence buffer = "";
    private int startOffset = 0;
    private int endOffset = 0;
    private int currentOffset = 0;
    private int state = 0;
    
    @Override
    public void start(@NotNull CharSequence buffer, int startOffset, int endOffset, int initialState) {
        this.buffer = buffer;
        this.startOffset = startOffset;
        this.endOffset = endOffset;
        this.currentOffset = startOffset;
        this.state = initialState;
    }
    
    @Override
    public int getState() {
        return state;
    }
    
    @Override
    public @Nullable IElementType getTokenType() {
        if (currentOffset >= endOffset) return null;
        
        char ch = buffer.charAt(currentOffset);
        if (Character.isWhitespace(ch)) {
            return FrugalTokenTypes.WHITE_SPACE;
        } else if (Character.isLetter(ch)) {
            return FrugalTokenTypes.IDENTIFIER;
        } else if (Character.isDigit(ch)) {
            return FrugalTokenTypes.NUMBER;
        } else if (ch == '"') {
            return FrugalTokenTypes.STRING;
        } else if (ch == '/' && currentOffset + 1 < endOffset && buffer.charAt(currentOffset + 1) == '/') {
            return FrugalTokenTypes.COMMENT;
        } else {
            return FrugalTokenTypes.SYMBOL;
        }
    }
    
    @Override
    public int getTokenStart() {
        return currentOffset;
    }
    
    @Override
    public int getTokenEnd() {
        if (currentOffset >= endOffset) return endOffset;
        
        char ch = buffer.charAt(currentOffset);
        int end = currentOffset + 1;
        
        if (Character.isWhitespace(ch)) {
            while (end < endOffset && Character.isWhitespace(buffer.charAt(end))) {
                end++;
            }
        } else if (Character.isLetter(ch)) {
            while (end < endOffset && (Character.isLetterOrDigit(buffer.charAt(end)) || buffer.charAt(end) == '_')) {
                end++;
            }
        } else if (Character.isDigit(ch)) {
            while (end < endOffset && Character.isDigit(buffer.charAt(end))) {
                end++;
            }
        } else if (ch == '"') {
            while (end < endOffset && buffer.charAt(end) != '"') {
                if (buffer.charAt(end) == '\\') end++;
                end++;
            }
            if (end < endOffset) end++;
        } else if (ch == '/' && currentOffset + 1 < endOffset && buffer.charAt(currentOffset + 1) == '/') {
            while (end < endOffset && buffer.charAt(end) != '\n') {
                end++;
            }
        }
        
        return end;
    }
    
    @Override
    public void advance() {
        currentOffset = getTokenEnd();
    }
    
    @Override
    public @NotNull CharSequence getBufferSequence() {
        return buffer;
    }
    
    @Override
    public int getBufferEnd() {
        return endOffset;
    }
}

File: src/main/kotlin/dev/strawn/frugal/FrugalLexer.kt
package dev.strawn.frugal

import com.intellij.lexer.LexerBase
import com.intellij.psi.tree.IElementType

class FrugalLexer : LexerBase() {
    
    private var buffer: CharSequence = ""
    private var startOffset: Int = 0
    private var endOffset: Int = 0
    private var currentOffset: Int = 0
    private var state: Int = 0
    
    override fun start(buffer: CharSequence, startOffset: Int, endOffset: Int, initialState: Int) {
        this.buffer = buffer
        this.startOffset = startOffset
        this.endOffset = endOffset
        this.currentOffset = startOffset
        this.state = initialState
    }
    
    override fun getState(): Int = state
    
    override fun getTokenType(): IElementType? {
        if (currentOffset >= endOffset) return null
        
        val char = buffer[currentOffset]
        return when {
            char.isWhitespace() -> FrugalTokenTypes.WHITE_SPACE
            char.isLetter() -> FrugalTokenTypes.IDENTIFIER
            char.isDigit() -> FrugalTokenTypes.NUMBER
            char == '"' -> FrugalTokenTypes.STRING
            char == '/' && currentOffset + 1 < endOffset && buffer[currentOffset + 1] == '/' -> FrugalTokenTypes.COMMENT
            else -> FrugalTokenTypes.SYMBOL
        }
    }
    
    override fun getTokenStart(): Int = currentOffset
    
    override fun getTokenEnd(): Int {
        if (currentOffset >= endOffset) return endOffset
        
        val char = buffer[currentOffset]
        var end = currentOffset + 1
        
        when {
            char.isWhitespace() -> {
                while (end < endOffset && buffer[end].isWhitespace()) {
                    end++
                }
            }
            char.isLetter() -> {
                while (end < endOffset && (buffer[end].isLetterOrDigit() || buffer[end] == '_')) {
                    end++
                }
            }
            char.isDigit() -> {
                while (end < endOffset && buffer[end].isDigit()) {
                    end++
                }
            }
            char == '"' -> {
                while (end < endOffset && buffer[end] != '"') {
                    if (buffer[end] == '\\') end++
                    end++
                }
                if (end < endOffset) end++
            }
            char == '/' && currentOffset + 1 < endOffset && buffer[currentOffset + 1] == '/' -> {
                while (end < endOffset && buffer[end] != '\n') {
                    end++
                }
            }
        }
        
        return end
    }
    
    override fun advance() {
        currentOffset = tokenEnd
    }
    
    override fun getBufferSequence(): CharSequence = buffer
    
    override fun getBufferEnd(): Int = endOffset
}